### ISSUE TITLE
Question about the file locking and the /f option in mirror.exe

### ISSUE THREAD

**Issue by yin19941005** [dokan-dev/dokany#423]
# Environment
 * Windows version:  Windows Server 2012 R2
 * Processor architecture:  amd64 with multiple cores
 * Dokany version: 1.0.2.1000 release / snapshot 1.0.1.1017
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
(This is a multi-thread issue, not guarantee reproduce on each time but in limited trial you can reproduce it)
  - [x] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [ ] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

The 1.0.2.1000 release should be latest version for master branch, so, I didn't test for the snapshot.

# Description
Race condition had occur on Windows Servers sharing the dokan drive and cause data corruption. The read and write operation both are under the race condition and could cause the data corruption. 

### Steps to reproduce:
**CAUTION: MAKE SURE THE SERVER GET MULTIPLE CORE, SINGLE CORE MACHINE CANNOT REPRODUCE THIS ISSUE!!**

1. Mount the drive with the mirror.exe on the Server
2. Sharing out the dokan drive by Windows build-in function on the Server
3. Client connect to the Server to access the dokan shared drive
4. Client Copying some files into the dokan shared drive (I am using image files)

The read operation and write operation are causing the data corruption independently. Sometime the file copied in the Server is normal but open it up on the Client will cause data corruption, and vice versa. And of course, they can happen in the same time.

I had also tested for the option `/t 1` to force the mirror only running on single thread and issue disappear. The **data corruption will not happen at all under `/t 1` option**.

I further tested on dokany version 1.0.1.1017 and the problem is still persist. I rollback to this version is because the 1.0.2.1000 release had applied some fix on locking and I am trying the older version to test the issue is related or not. And it seems the bug fix on the locking is not the reason causing this issue because **I am still able to reproduce this issue on version 1.0.1.1017**.

I had also tested for Windows 7 and Windows 10. They rarely happen the race condition on shared dokan drive and cause the data corruption. For your reference, I can only reproduce once when I keep testing on Windows 7 and Windows 10 for 30 minutes.

Command used to mount : `mirror.exe /r E:\t1 /l M /d /s`

# Logs
The mirror debug log:
[Test002.log.txt](https://github.com/dokan-dev/dokany/files/731486/Test002.log.txt)
I am not sure which parts will be useful for race condition issue, so, this is the full version.

![multi-thread-command-promt-001](https://cloud.githubusercontent.com/assets/23186757/22318818/f5f915e4-e3b8-11e6-8134-5f90d553879a.png)

![multi-thread-issue-001](https://cloud.githubusercontent.com/assets/23186757/22318819/f7cb4aae-e3b8-11e6-84d3-e43fb166f41c.png)

**Comment by Liryna** [comment]
Hi @yin19941005 ,

This is a normal behavior to expect. All read and write dokan operation have to be thread safe like explained in the documentation.
https://dokan-dev.github.io/dokany-doc/html/struct_d_o_k_a_n___o_p_e_r_a_t_i_o_n_s.html#ab1aa9dab25dcb38903e0790c36b87cce

You need to add a lock/mutex in your implementation to be sure that the handle/data are safely manipulated.

**Comment by yin19941005** [comment]
Hi @Liryna ,

Thank you for your kindly explanation!! I got a bit confused for new option `DOKAN_OPTION_FILELOCK_USER_MODE` (/f option). 

Do you mean that I should either implementing the lock/mutex in the read/write operation or enable the `/f` option and implementing the lock function in the mirror (& call the mirrorLockFile in read/write)?

**Comment by marinkobabic** [comment]
Can the issue be reproduced with one file and putting in some delays during
write operations?

You are going to write a bigger file and an client reads the same file
already during a longer taking write operations which is still not
completed for the whole  file. Only 60% of the file have been written at
this point of read.

Am 27.01.2017 03:07 schrieb "yin19941005" <notifications@github.com>:

**Comment by Liryna** [comment]
Hi @yin19941005,

Yes, you should have lock/mutex on the read/write operation in any case to support OVERLAPPED read/write (maybe this flag is set in your case ?)
https://msdn.microsoft.com/en-us/library/windows/desktop/aa365683(v=vs.85).aspx

**Comment by yin19941005** [comment]
Hi @Liryna ,

Thank you for your kindly explanation! I had confirmed the OVERLAPPED flag is not set in my case but the issue can be fixed by adding Mutex on the read/write operation.

I think the problem is come from the CIFS share may have a buffer and this explicit the race condition on the read/write operation even the OVERLAPPED flag is not set. I have a VM which usually can reproduce the issue on each test of copying the files. After adding a Mutex for both read and write function in mirror, the data corruption issue gone.

<br>

@marinkobabic ,

It didn't seems the issue is caused by reading the incomplete write files. I had checked the log files and it doesn't trying to read the file before the last write operation complete to the same file. I think the cause of this issue is CIFS would causing the read/write operation similar with the case as OVERLAPPED is set (maybe because the buffer?). Is my fault that didn't confirm the document carefully.

<br>

Thank you for you guys' kindly explanation again!!
